staging/src/k8s.io/apiserver/pkg/storage/etcd3/decoder.go

```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd3
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"k8s.io/apimachinery/pkg/conversion"
+	"k8s.io/apimachinery/pkg/runtime"
+	endpointsrequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/storage"
+
+	"k8s.io/klog/v2"
+)
+
+// NewDefaultDecoder returns the default decoder for etcd3 store
+func NewDefaultDecoder() Decoder { return &defaultDecoder{} }
+
+// Decoder is used by the etcd storage implementation to decode
+// transformed data from the storage into an object
+type Decoder interface {
+	// Decode decodes value of bytes into object. It will also
+	// set the object resource version to rev.
+	// On success, objPtr would be set to the object.
+	Decode(codec runtime.Codec, versioner storage.Versioner, value []byte, objPtr runtime.Object, rev int64) error
```

Comment:
The fact that we pass `codec` as a parameter to `Decoder` seems a bit weird to me (same for DecodeListItems below).

Shouldn't `codec` be passed on initialization of Decoder (i.e. codec as a field of `defaultDecoder struct`) ?

Reply:
yes, in contrast to the request scoped parameters, both `codec` and `versioner` are out of band and shared, so they can be promoted to member variables. Do you want me to add `versioner` as a field of `defaultDecoder` as well?

Reply:
okay, done, please take a look when you have a moment